tc/autotuner/options_cache.h

```diff
@@ -0,0 +1,179 @@
+
+namespace tc {
+namespace autotune {
+
+/**
+ * A key in the options cache is exactly the content of the underlying proto in
+ * tc/proto/compcache.proto. It provides simple conversions and equality
+ * operator. Additionally we provide a hash function to allow it to be a key
+ * in a hash map.
+ */
+struct OptionsCacheKey {
+  lang::CanonicalTcString id;
+  std::vector<TensorInfo> inputs;
+  std::vector<TensorInfo> outputs;
+  std::string deviceStr;
+
+  inline bool operator==(const OptionsCacheKey& other) const;
+  inline OptionsCacheKeyProto toProtobuf() const;
+  static inline OptionsCacheKey fromProtobuf(const OptionsCacheKeyProto& proto);
+};
+
+struct OptionsCacheKeyHash {
+  inline std::size_t operator()(const OptionsCacheKey& k) const;
+};
+
+/**
+ * A value in the options cache is exactly the content of the underlying proto
+ * in tc/proto/compcache.proto. An OptionsCacheValue is templated by the Backend
+ * and contains an object of the proper MappingOptions type as well as the
+ * runtimes of subsequent runs.
+ */
+template <typename Backend>
+struct OptionsCacheValue {
+  typename Backend::OptionsCacheValueProtoType toProtobuf() const;
+  static OptionsCacheValue<Backend> fromProtobuf(
+      const typename Backend::OptionsCacheValueProtoType& proto);
+
+  std::vector<Duration> runtimes;
+  typename Backend::MappingOptionsType mappingOptions;
+};
+
+/**
+ * An Options cache is a simple abstraction around an unordered_multimap of
+ * protobuf-backed key value pairs. It has an underlying store object and
+ * provide simple functions to load/store from proto, record, prune and
+ * extract topK values ordered by runtime.
+ * An OptionsCache is templated by the backend type because the values stored
+ * are backend-dependent.
+ */
+template <typename Backend>
+struct OptionsCache {
+ public:
+  using KeyType = OptionsCacheKey;
+  using ValueType = OptionsCacheValue<Backend>;
+  using MultiMapType = std::unordered_multimap<
+      OptionsCacheKey,
+      OptionsCacheValue<Backend>,
+      OptionsCacheKeyHash>;
+
+  OptionsCache<Backend>();
+
+  /// Clears the content of the cache and resets the counters
+  void clear();
+
+  /// \return the number of values for a particular key
+  size_t count(const OptionsCacheKey& key) const;
+
+  //// \return the number of element in the cache
+  size_t size() const;
+
+  /// Collects the keys in the cache
+  /// \return an unordered_set of keys
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> getKeys() const;
+
+  /// Loads in place from proto file. Calls fromProto which can insert
+  /// duplicates, so be sure your cache is cleared if you don't want those
+  void loadCacheFromFile(const std::string& filename);
+
+  /// Stores to a proto file at the specified location
+  void storeCacheToFile(const std::string& filename) const;
+
+  /// Saves a new runtime.
+  /// If the key does not exist, a new entry is inserted.
+  /// If the key exists, a linear search is performed on options.
```

Comment:
This describes _how_ the function works (linear search), rather than _what_ it does.
Is this necessary here?